stan/math/opencl/prim/sum.hpp

```diff
@@ -0,0 +1,37 @@
+#ifndef STAN_MATH_OPENCL_PRIM_SUM_HPP
+#define STAN_MATH_OPENCL_PRIM_SUM_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/sum.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+
+namespace stan {
+namespace math {
+
+/**
+ * Calculates sum of given kernel generator expression.
+ * @tparam T type of the expression
+ * @param m expression to sum
+ * @return sum of given expression
+ */
+template <typename T, require_all_kernel_expressions_t<T>* = nullptr>
+value_type_t<T> sum(const T& m) {
+  matrix_cl<value_type_t<T>> res;
+  if (m.rows() == 1) {
+    // using rowwise_sum would run just 1 thread
+    res = colwise_sum(transpose(m));
+  } else if (m.cols() == 1) {
+    res = colwise_sum(m);
+  } else {
+    res = colwise_sum(rowwise_sum(m));
+  }
+  return sum(from_matrix_cl<Eigen::Dynamic, 1>(res));
+}
```

Comment:
[optional] It would be a bit harder to read but rvo can be used here if instead of using `res` there are multiple returns like below

https://godbolt.org/z/4q65zf

Reply:
Also is there ever an instance we'd want to keep this single value over on the gpu (i.e. summing stuff on the gpu)? If so then maybe this function should return back a `matrix_cl`. Then at the call site we would do we can do the `from_matrix_cl`

Reply:
Res is not what is returned - it is copied to host and summed before returning.

> Also is there ever an instance we'd want to keep this single value over on the gpu

There probably exist use cases where this would be the optimal thing to do. However, I think what is in this PR is not significantly slower. Also if we want to do the optimal thing it is probably not to sum in a separate kernel, but instead to directly use `res` in the next kernel an sum it as a part of that kernel. That is getting a bit complicated and probably not necessary. So what we have here should be good enough.

Reply:
imo it's not about efficiency, it's for API consistency since for all the other ops we stay on the gpu. I understand that this would get wrapped in `scalar_<>` anyway, but it feels odd to have one function that returns something not on the gpu

Reply:
I could also argue that the way I implemented it is also for consistency. Namely between overloads of the `sum` function. They all return a scalar. This is actually needed in https://github.com/bstatcomp/math/blob/8116932357a8492b754f91436429d978a6982379/stan/math/prim/functor/broadcast_array.hpp#L32 (this is used in `operands_and_partials7 and can be called with a `matrix_cl`).

Reply:
lol this function didn't exist when that code went in tho'! For the rest of the API we make it clear when we cross the boundary from OpenCL and back. I'd rather have an overload for that operator when passing stuff back from the GPU. Are there other places we would want to be swapping back and forth implicitly from a `matrix_cl` to an Eigen matrix (or any boundary crossing)?